contracts/upkeeps/Cron.sol

```diff
@@ -0,0 +1,287 @@
+      .decode(performData, (uint256, uint256, address, bytes));
+    validate(id, tickTime, target, handler);
+    s_lastRuns[id] = block.timestamp;
+    (bool success, bytes memory payload) = target.call(handler);
+    if (!success) {
+      revert CallFailed(id, getRevertMsg(payload));
+    }
+    emit CronJobExecuted(id, block.timestamp);
+  }
+
+  /**
+   * @notice Creates a cron job from the given encoded spec
+   * @param target the destination contract of a cron job
+   * @param handler the function signature on the target contract to call
+   * @param encodedCronSpec abi encoding of a cron spec
+   */
+  function createCronJobFromEncodedSpec(
+    address target,
+    bytes memory handler,
+    bytes memory encodedCronSpec
+  ) external {
+    Spec memory spec = abi.decode(encodedCronSpec, (Spec));
+    createCronJobFromSpec(target, handler, spec);
+  }
+
+  /**
+   * @notice Deletes the cron job matching the provided id. Reverts if
+   * the id is not found.
+   * @param id the id of the cron job to delete
+   */
+  function deleteCronJob(uint256 id) external onlyOwner {
+    uint256 oldCount = s_activeCronJobIDs.length;
+    uint256[] memory newActiveCronJobIDs = new uint256[](oldCount - 1);
+    uint256 newIdx;
+    uint256 oldId;
+    for (uint256 idx = 0; idx < oldCount; idx++) {
+      oldId = s_activeCronJobIDs[idx];
+      if (oldId == id) {
+        continue;
+      }
+      if (newIdx == oldCount - 1) {
+        revert CronJobIDNotFound(id);
+      }
+      newActiveCronJobIDs[newIdx] = oldId;
+      newIdx++;
+    }
+    s_activeCronJobIDs = newActiveCronJobIDs;
+    delete s_lastRuns[id];
+    delete s_specs[id];
+    delete s_targets[id];
+    delete s_handlers[id];
+    delete s_handlerSignatures[id];
+    emit CronJobDeleted(id);
+  }
+
+  /**
+   * @notice Pauses the contract, which prevents executing performUpkeep
+   */
+  function pause() external onlyOwner {
+    _pause();
+  }
+
+  /**
+   * @notice Unpauses the contract
+   */
+  function unpause() external onlyOwner {
+    _unpause();
+  }
+
+  /**
+   * @notice Get the id of an eligible cron job
+   * @return upkeepNeeded signals if upkeep is needed, performData is an abi encoding
+   * of the id and "next tick" of the elligible cron job
+   */
+  function checkUpkeep(bytes calldata)
+    external
+    view
+    override
+    whenNotPaused
+    returns (bool, bytes memory)
+  {
+    // DEV: start at a random spot in the list so that checks are
+    // spread evenly among cron jobs
+    uint256 idx;
+    uint256 id;
+    uint256 lastTick;
+    uint256 numCrons = s_activeCronJobIDs.length;
+    uint256 startIdx = block.number % numCrons;
+    for (idx = startIdx; idx < numCrons; idx++) {
+      id = s_activeCronJobIDs[idx];
+      lastTick = s_specs[id].lastTick();
+      if (lastTick > s_lastRuns[id]) {
+        return (true, abi.encode(id, lastTick, s_targets[id], s_handlers[id]));
+      }
+    }
+    for (idx = 0; idx < startIdx; idx++) {
```

Comment:
> Or is there a reason to keep both?

Here is what I'm trying to avoid... if we only encode / decode to structs, when the output looks like this:
![image](https://user-images.githubusercontent.com/14809513/140803229-6b9e6cea-323b-442b-8051-327ec2df3f4e.png)
and a user has to manually convert that to the correctly formatted 5 * 32 struct...
```
[[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]]
```
(I still haven't gotten passing a struct to work in remix)

Vs encoding the struct...
![image](https://user-images.githubusercontent.com/14809513/140804586-c58bf572-3756-4787-89f9-4a6b72519f2e.png)
Here, you just c/p the payload into `createCronJobFromEncodedSpec()` and it just "works". No need to fuss with properly formatting the tuple.

Granted, if folks do their deployments programmatically, it will be cheaper to use the struct, but if users are doing contract interactions with remix or Etherscan, it's going to be a PITA to use the tuple.

Reply:
FYI I did some gas testing on the difference, and it's about 22K gas cheaper to use the structs.
```
···················|································|·············|·············|··············|
|  Contract        ·  Method                        ·  Min        ·  Max        ·  Avg         ·
···················|································|·············|·············|··············|
|  CronTestHelper  ·  createCronJobFromEncodedSpec  ·     265131  ·     384711  ·      322055  ·
···················|································|·············|·············|··············|
|  CronTestHelper  ·  createCronJobFromSpec         ·     259782  ·     362262  ·      300827  ·
···················|································|·············|·············|··············|
```

Reply:
>  it would be nice to DRY them up and extract...
Done :+1: